storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoArray.java

```diff
@@ -639,13 +639,44 @@ public boolean nodeIsViableForHead(ProtoNode node) {
                   .isGreaterThanOrEqualTo(currentEpoch);
     }
 
-    final UInt64 finalizedSlot = spec.computeStartSlotAtEpoch(finalizedCheckpoint.getEpoch());
     final boolean correctFinalized =
         node.getFinalizedCheckpoint().getEpoch().equals(initialEpoch)
-            || hasAncestorAtSlot(node, finalizedSlot, finalizedCheckpoint.getRoot());
+            || isFinalizedRootOrDescendant(node);
     return correctJustified && correctFinalized;
   }
 
+  private boolean isFinalizedRootOrDescendant(final ProtoNode node) {
+    final UInt64 finalizedEpoch = finalizedCheckpoint.getEpoch();
+    final Bytes32 finalizedRoot = finalizedCheckpoint.getRoot();
+
+    final Checkpoint nodeFinalizedCheckpoint = node.getFinalizedCheckpoint();
+    if (nodeFinalizedCheckpoint.getEpoch().equals(finalizedEpoch)
+        && nodeFinalizedCheckpoint.getRoot().equals(finalizedRoot)) {
+      return true;
+    }
+
+    final Checkpoint nodeJustifiedCheckpoint = node.getJustifiedCheckpoint();
+    if (nodeJustifiedCheckpoint.getEpoch().equals(finalizedEpoch)
+        && nodeJustifiedCheckpoint.getRoot().equals(finalizedRoot)) {
+      return true;
+    }
+
+    final Checkpoint nodeUnrealizedFinalizedCheckpoint = node.getUnrealizedFinalizedCheckpoint();
+    if (nodeUnrealizedFinalizedCheckpoint.getEpoch().equals(finalizedEpoch)
+        && nodeUnrealizedFinalizedCheckpoint.getRoot().equals(finalizedRoot)) {
+      return true;
+    }
+
+    final Checkpoint nodeUnrealizedJustifiedCheckpoint = node.getUnrealizedJustifiedCheckpoint();
+    if (nodeUnrealizedJustifiedCheckpoint.getEpoch().equals(finalizedEpoch)
+        && nodeUnrealizedJustifiedCheckpoint.getRoot().equals(finalizedRoot)) {
+      return true;
+    }
+
+    final UInt64 finalizedSlot = spec.computeStartSlotAtEpoch(finalizedCheckpoint.getEpoch());
+    return hasAncestorAtSlot(node, finalizedSlot, finalizedRoot);
```

Comment:
im not sure how we can ever get here but fail the check at 653?

Reply:
ok i figured this out, its complicated and we can talk about why if we need to but i'll resolve.

Reply:
I agree with the @rolfyone 's doubts that there might be no cases where we don't match the 4 conditions and we ending up returning `true` via `hasAncestorAtSlot`.

But I tried to hardcode `return false` at the end run fork_choice ref tests and fails quite a lot of them. So there are definitely cases!